assets/scripts/app/load_resources.js

```diff
@@ -39,38 +17,132 @@ const IMAGES_TO_BE_LOADED = [
+window.imagesToBeLoaded = 1
+
+// Load everything
+loadImages()
+loadSVGs()
+
+// When everything is loaded...
+Promise.all(loading)
+  .then(function () {
+    // Export to window (LEGACY)
+    window.images = images
+    window.imagesToBeLoaded = 0
+
+    // Also legacy, TODO: replace with promise
+    _checkIfEverythingIsLoaded()
+  })
 
 function loadImages () {
-  imagesToBeLoaded = IMAGES_TO_BE_LOADED.length
+  loadingEl.max += IMAGES_TO_BE_LOADED.length
 
-  for (var i in IMAGES_TO_BE_LOADED) {
-    var url = IMAGES_TO_BE_LOADED[i]
+  for (let url of IMAGES_TO_BE_LOADED) {
+    loading.push(window.fetch(url + '?v' + TILESET_IMAGE_VERSION)
+      .then(function (response) {
+        return response.blob()
+      })
+      .then(function (blob) {
+        const img = new Image()
+        img.src = window.URL.createObjectURL(blob)
 
-    images[url] = document.createElement('img')
-    images[url].addEventListener('load', onImageLoaded)
-    images[url].src = url + '?v' + TILESET_IMAGE_VERSION
-  }
+        // Store on the global images object, using the url as the key
+        images[url] = img
 
-  document.querySelector('#loading-progress').value = 0
-  document.querySelector('#loading-progress').max = imagesToBeLoaded + 5
+        loadingEl.value++
+      })
+      .catch(function (error) {
+        console.error('loading image error', error)
+      }))
+  }
 }
 
-function onImageLoaded () {
-  imagesToBeLoaded--
-  document.querySelector('#loading-progress').value++
+function loadSVGs () {
+  loadingEl.max += SVGS_TO_BE_LOADED.length
+
+  for (let url of SVGS_TO_BE_LOADED) {
+    loading.push(window.fetch(url)
+      .then(function (response) {
+        return response.text()
+      })
+      .then(function (response) {
+        SVGStagingEl.innerHTML += response
+
+        // ctx.drawImage() can only draw things that are images, so you can't draw
+        // an SVG directly. You also can't <use> a symbol reference from inside an
+        // image tag. So we have to create an image using a reconstructed SVG as a
+        // data-URI. Here, let's cache all the artwork svgs as image elements for
+        // later rendering to canvas
+        let svgEls = SVGStagingEl.querySelectorAll('symbol')
+
+        for (let svg of svgEls) {
+          // Only cache segment illustrations, don't need to cache icons
+          if (svg.id.indexOf('image-') === 0) {
+            // Simplify id, removing namespace prefix
+            const id = svg.id.replace(/^image-/, '')
+
+            // Get details of the SVG so we can reconstruct an image element
+            const svgViewbox = svg.getAttribute('viewBox')
+            let svgInternals = svg.innerHTML
+
+            // innerHTML is not an available property for SVG elements in IE / Edge
+            // so if turns to be undefined, we use this alternate method below
+            // to clone the SVG's nodes to a dummy element and then grab the
+            // innerHTML out of that one.
```

Comment:
Not sure what level of browser compatibility you require, I’ve found the `XMLSerializer` to be simple and reliable (IE 9+, everything else from way back):

`(new XMLSerializer()).serializeToString(svgNode)`